File: app/src/main/java/com/example/andreip/myapplication/app/RetroFitAdapter.java
package com.example.andreip.myapplication.app;

import android.util.Log;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetroFitAdapter {

    private Retrofit retrofit;
    private ArrayList<Recipe> recipeArrayList = null;
    private MainActivity mainActivity;

    public RetroFitAdapter(MainActivity mainActivity) {
        this.mainActivity = mainActivity;
    }

    public ArrayList<Recipe> getRecipeArrayList() {
        return recipeArrayList;
    }

    public void setRecipeArrayList(ArrayList<Recipe> recipeArrayList) {
        this.recipeArrayList = recipeArrayList;
    }

    public void filterRecipeArrayList(ArrayList<Recipe> recipeArrayList, String text) {
        for (Recipe recipe : recipeArrayList)
            if (!recipe.getTitle().contains(text)) {
                recipeArrayList.remove(recipe);
            }
        setRecipeArrayList(recipeArrayList);
    }

    public static OkHttpClient getClient() {

        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        return new OkHttpClient.Builder()
                .addInterceptor(interceptor)
                .build();
    }

    private void initializeRetroFit() {
        retrofit = new Retrofit.Builder()
                .baseUrl("http://www.recipepuppy.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .client(getClient())
                .build();
    }

    public void getRecipes(final String s) {
        initializeRetroFit();
        List<String> veggies = new ArrayList<>();
        veggies.add("onions");
        veggies.add("garlic");
        RecipePuppyAPI service = retrofit.create(RecipePuppyAPI.class);

        Call<RecipeResponse> request = service.getRecipies(veggies, "omelet");
        request.enqueue(new Callback<RecipeResponse>() {
            @Override
            public void onResponse(Call<RecipeResponse> call, Response<RecipeResponse> response) {
                if (response.isSuccessful()) {
                    if (response.body() != null) {
                        if (response.body().getRecipes().equals(null))
                            Toast.makeText(mainActivity.getApplicationContext(), "Empty Response", Toast.LENGTH_SHORT);
                        Log.d("TITLE", response.body().getRecipes().get(0).getIngredients());
                        setRecipeArrayList(response.body().getRecipes());

                        if (!s.equalsIgnoreCase(""))
                            filterRecipeArrayList(getRecipeArrayList(), s);

                        ((MyAdapter) mainActivity.getRecyclerView().getAdapter())
                                .addAllItems(getRecipeArrayList());
                    }
                } else {
                    Log.d("FAIL", response.code() + " " + response.message());
                }
            }

            @Override
            public void onFailure(Call<RecipeResponse> call, Throwable t) {
                Log.e("FAIL", t.getMessage() + " ", t);
            }
        });

    }
}

File: app/src/main/java/com/example/andreip/myapplication/app/RetroFitAdapter.kt
package com.example.andreip.myapplication.app

import android.util.Log

import java.util.ArrayList

import SQLData.RecipeDb
import SQLData.WordRepository
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class RetroFitAdapter(private val mainActivity: MainActivity) {

    private var retrofit: Retrofit? = null
    var recipeArrayList: List<Recipe> = ArrayList<Recipe> ()

    fun toRecipe(recipeDbList: List<RecipeDb>): ArrayList<Recipe>{
        var recipeArrayList = ArrayList<Recipe>()
        for (recipeDb in recipeDbList)
            recipeArrayList.add(Recipe(recipeDb))

        return recipeArrayList
    }

    fun filterRecipeArrayList(adapter: MyAdapter, recipeArrayList: List<Recipe>?, text: String) {

        val filterRecipeList = ArrayList<Recipe>()

        for (recipe in recipeArrayList!!)
            if (recipe.getTitle()!!.contains(text)) {
                filterRecipeList.add(recipe)
            }

        this.recipeArrayList = filterRecipeList
        adapter.refreshAllItems(filterRecipeList)
    }

    private fun initializeRetroFit() {
        retrofit = Retrofit.Builder()
                .baseUrl("http://www.recipepuppy.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build()
    }


    fun getRecipes(s: String, wordRepository: WordRepository) {
        initializeRetroFit()
        val veggies = ArrayList<String>()
        veggies.add("onions")
        veggies.add("garlic")
        val service = retrofit!!.create(RecipePuppyAPI::class.java)

        val request = service.getRecipies(veggies, "omelet")
        request.enqueue(object : Callback<RecipeResponse> {
            override fun onResponse(call: Call<RecipeResponse>, response: Response<RecipeResponse>) {
                if (response.isSuccessful) {
                    if (response.body() != null) {
                        Log.d("TITLE", response.body()!!.recipes!![0].getIngredients())
                        recipeArrayList = response.body()!!.recipes

                        if (!s.equals("", ignoreCase = true))
                            filterRecipeArrayList(mainActivity.recyclerView.adapter as MyAdapter, recipeArrayList, s)
                        else
                            (mainActivity.recyclerView.adapter as MyAdapter).addAllItems(recipeArrayList)

                        Log.d("TITLE_DB", wordRepository.getValueAllWords().toString() + " ")

                        wordRepository.insertAllRecipes(recipeArrayList)
                    }
                } else {
                    recipeArrayList = toRecipe(wordRepository.getValueAllWords())


                    if (!s.equals("", ignoreCase = true))
                        filterRecipeArrayList(mainActivity.recyclerView.adapter as MyAdapter, recipeArrayList, s)
                    else
                        (mainActivity.recyclerView.adapter as MyAdapter).addAllItems(recipeArrayList)
                }
            }

            override fun onFailure(call: Call<RecipeResponse>, t: Throwable) {

                recipeArrayList = toRecipe(wordRepository.getValueAllWords())
                Log.e("FAIL", t.message + " ", t)

                if (!s.equals("", ignoreCase = true))
                    filterRecipeArrayList(mainActivity.recyclerView.adapter as MyAdapter, recipeArrayList, s)
                else
                    (mainActivity.recyclerView.adapter as MyAdapter).addAllItems(recipeArrayList)
            }
        })

    }

    companion object {

        val client: OkHttpClient
            get() {

                val interceptor = HttpLoggingInterceptor()
                interceptor.level = HttpLoggingInterceptor.Level.BODY

                return OkHttpClient.Builder()
                        .addInterceptor(interceptor)
                        .build()
            }
    }
}


